test/WebSites/RazorPageExecutionInstrumentationWebSite/Startup.cs

```diff
@@ -22,24 +20,28 @@ public void ConfigureServices(IServiceCollection services)
 
             // Add MVC services to the services container.
             services.AddMvc();
-
-            // Make instrumentation data available in views.
-            services.AddScoped<TestPageExecutionListenerFeature, TestPageExecutionListenerFeature>();
-            services.AddScoped<IHoldInstrumentationData>(
-                provider => provider.GetRequiredService<TestPageExecutionListenerFeature>().Holder);
         }
 
         public void Configure(IApplicationBuilder app)
         {
+            var listener = new RazorPageDiagnosticListener();
+            var diagnosticSource = app.ApplicationServices.GetRequiredService<DiagnosticListener>();
+            diagnosticSource.SubscribeWithAdapter(listener);
+
             app.UseCultureReplacer();
 
-            // Execute views with instrumentation enabled.
-            app.Use((HttpContext context, Func<Task> next) =>
+            app.Use(async (context, next) =>
             {
-                var listenerFeature = context.RequestServices.GetRequiredService<TestPageExecutionListenerFeature>();
-                context.Features.Set<IPageExecutionListenerFeature>(listenerFeature);
-
-                return next();
+                // Add MVC to the request pipeline
+                await next();
+
+                using (var writer = new HttpResponseStreamWriter(context.Response.Body, Encoding.UTF8))
```

Comment:
Suggestion from @lodejard: Give MVC `TextWriter.Null` and have `RazorPageDiagnosticListener` write directly to the response as the events occur. No reason to cache the events anywhere.